src/openlcb/StreamSender.hxx

```diff
@@ -0,0 +1,555 @@
+    /// @param source_stream_id 8-bit stream ID to use on the this (the source)
+    /// side.
+    ///
+    /// @return *this for calling optional settings API commands.
+    ///
+    StreamSenderCan &start_stream(NodeHandle dst, uint8_t source_stream_id)
+    {
+        DASSERT(state_ == IDLE);
+        state_ = STARTED;
+        dst_ = dst;
+        totalByteCount_ = 0;
+        localStreamId_ = source_stream_id;
+        dstStreamId_ = 0xFF;
+        HASSERT(sleeping_ == false);
+        HASSERT(requestClose_ == 0);
+        requestInit_ = true;
+        trigger();
+        streamFlags_ = 0;
+        streamAdditionalFlags_ = 0;
+        streamWindowSize_ = StreamDefs::MAX_PAYLOAD;
+        streamWindowRemaining_ = 0;
+        errorCode_ = 0;
+        return *this;
+    }
+
+    /// Closes the stream when all the bytes are transferred.
+    void close_stream()
+    {
+        requestClose_ = true;
+        trigger();
+    }
+
+    /// Specifies what the source should propose as window size to the
+    /// destination. May be called only after start_stream.
+    ///
+    /// @param window_size in bytes, what should we propose in the stream
+    /// initiate call
+    ///
+    StreamSenderCan &set_proposed_window_size(uint16_t window_size)
+    {
+        HASSERT(state_ == STARTED);
+        streamWindowSize_ = window_size;
+        return *this;
+    }
+
+    /// Specifies the Stream UID to send in the stream initiate request. May be
+    /// called only after start_stream. This function must be used if opening
+    /// an unannounced stream to a destination.
+    ///
+    /// @param stream_uid a valid 6-byte stream identifier.
+    ///
+    StreamSenderCan &set_stream_uid(NodeID stream_uid)
+    {
+        HASSERT(state_ == STARTED);
+        /// @todo implement opening unannounced streams.
+        return *this;
+    }
+
+    /// Sets the stream sender to be available for reuse after a stream has
+    /// been closed or reached error.
+    void clear()
+    {
+        if (state_ == STATE_ERROR || state_ == CLOSING)
+        {
+            state_ = IDLE;
+        }
+    }
+
+#ifdef GTEST
+    /// Requests to exit any timed operation.
+    /// @return true if a timer was woken up.
+    bool shutdown()
+    {
+        if (sleeping_)
+        {
+            timer_.trigger();
+            sleeping_ = false;
+            return true;
+        }
+        return false;
+    }
+#endif
+
+    /// @return the state of this stream sender.
+    StreamSenderState get_state()
+    {
+        return state_;
+    }
+
+    /// @return the error code if we got a rejection from the remote node.
+    uint16_t get_error()
+    {
+        return errorCode_;
+    }
+    
+    /// Start of state machine, called when a buffer of data to send arrives
+    /// from the application layer.
+    Action entry()
+    {
+        if (requestInit_)
+        {
+            /// @todo get the destination address somehow.
```

Comment:
Do we already have the destination from start_stream()?

Reply:
yes, this was an outdated todo. removed.